datasetinsights/stats/visualization/bbox3d_plot.py

```diff
@@ -0,0 +1,160 @@
+
+    if color is None:
+        color = [0, 255, 0, 255]
+
+    _cv2.line(image, fbl, ful, color, box_line_width)  # front left
+    _cv2.line(image, ful, fur, color, box_line_width)  # front top
+    _cv2.line(image, fbr, fur, color, box_line_width)  # front right
+    _cv2.line(image, fbl, fbr, color, box_line_width)  # front bottom
+
+    _cv2.line(image, bbl, bul, color, box_line_width)  # back left
+    _cv2.line(image, bul, bur, color, box_line_width)  # back top
+    _cv2.line(image, bbr, bur, color, box_line_width)  # back right
+    _cv2.line(image, bbl, bbr, color, box_line_width)  # back bottom
+
+    _cv2.line(image, ful, bul, color, box_line_width)  # top left
+    _cv2.line(image, fur, bur, color, box_line_width)  # top right
+    _cv2.line(image, fbl, bbl, color, box_line_width)  # bottom left
+    _cv2.line(image, fbr, bbr, color, box_line_width)  # bottom right
+
+
+def add_single_bbox3d_on_image(
+    image, box, proj, color=None, box_line_width=2,
+):
+    """" Add single 3D bounding box on a given image.
+
+    Args:
+        image (numpy array): a numpy array for an image
+        box (BBox3D): a 3D bounding box in camera's coordinate system
+        proj (numpy 2D array): camera's 3x3 projection matrix
+        color(tuple): RGBA color of the bounding box. Defaults to None. If
+        color = None the the tuple of [0, 255, 0, 255] (Green) will be used.
+        box_line_width (int): line width of the bounding boxes. Defaults to 2.
+    """
+    img_height, img_width, _ = image.shape
+
+    fll = box.back_left_bottom_pt
+    ful = box.back_left_top_pt
+    fur = box.back_right_top_pt
+    flr = box.back_right_bottom_pt
+
+    bll = box.front_left_bottom_pt
+    bul = box.front_left_top_pt
+    bur = box.front_right_top_pt
+    blr = box.front_right_bottom_pt
+
+    fll_raster = _project_pt_to_pixel_location(fll, proj, img_height, img_width)
+    ful_raster = _project_pt_to_pixel_location(ful, proj, img_height, img_width)
+    fur_raster = _project_pt_to_pixel_location(fur, proj, img_height, img_width)
+    flr_raster = _project_pt_to_pixel_location(flr, proj, img_height, img_width)
+    bll_raster = _project_pt_to_pixel_location(bll, proj, img_height, img_width)
+    bul_raster = _project_pt_to_pixel_location(bul, proj, img_height, img_width)
+    bur_raster = _project_pt_to_pixel_location(bur, proj, img_height, img_width)
+    blr_raster = _project_pt_to_pixel_location(blr, proj, img_height, img_width)
+
+    _add_single_bbox3d_on_image(
+        image,
+        fll_raster,
+        ful_raster,
+        fur_raster,
+        flr_raster,
+        bll_raster,
+        bul_raster,
+        bur_raster,
+        blr_raster,
+        color,
+        box_line_width,
+    )
+
+
+def _project_pt_to_pixel_location(pt, projection, img_height, img_width):
+    """ Projects a 3D coordinate into a pixel location.
+
+    Applies the passed in projection matrix to project a point from the camera's
```

Comment:
I am a little rusty on the math that project a point from 3D to 2D image space. Can you add a wiki links or some textbook reference that explain this projection?

Reply:
updated